### ISSUE TITLE
Canu runtime 

### ISSUE THREAD

**Issue by Mondlii** [marbl/canu#2001]
Hi, I am relatively new to genome assembly and I wanted to compare different assemblers on the ONT reads I have. I am assembling a 680Mb plant genome and I have about 260x coverage for the reads. So far I've run 3 different assemblers to completion with the data but I am struggling with Canu. I am running on the grid with 450GB memory and 28 cores and the first run exceeded the 21 day runtime limit. I am using version 2.1.1 of canu with the below command:  

```canu -p Canu_AllData_1000 -d /projects/AC_SP6/Canu_assembly genomeSize=680m -minReadLength=1000 -useGrid=False -nanopore /AC_SP6/PromethION.fa NB. Our server does not support useGrid=True```

Due to memory restrictions I've had to remove the folder so I no longer have the same log file, but I have started a new run using corrected reads and giving it the -assemble tag thinking it would be faster but it is now more than 8 days into the assembly and the run is stuck in the same process as the last time. the first few days of the "Starting 'utgovl' concurrent execution" were at a decent rate with more than 3 steps completing per few hours, but now it seems to be stuck at at most 2 steps per day. At this rate it will again exceed the 21 day limit cause it has more than 80 processes to run. The command used for this run is:

```canu -assemble -p Canu_AllData_1000 -d /projects/AC_SP6/Canu_assembly genomeSize=680m -minReadLength=1000 -useGrid=False -corrected -nanopore /AC_SP6/Cns_final.fa```

Please see the report below.
```
-- canu 2.1.1
--
-- CITATIONS
--
-- For 'standard' assemblies of PacBio or Nanopore reads:
--   Koren S, Walenz BP, Berlin K, Miller JR, Phillippy AM.
--   Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation.
--   Genome Res. 2017 May;27(5):722-736.
--   http://doi.org/10.1101/gr.215087.116
-- 
-- Read and contig alignments during correction and consensus use:
--   Šošic M, Šikic M.
--   Edlib: a C/C ++ library for fast, exact sequence alignment using edit distance.
--   Bioinformatics. 2017 May 1;33(9):1394-1395.
--   http://doi.org/10.1093/bioinformatics/btw753
-- 
-- Overlaps are generated using:
--   Berlin K, et al.
--   Assembling large genomes with single-molecule sequencing and locality-sensitive hashing.
--   Nat Biotechnol. 2015 Jun;33(6):623-30.
--   http://doi.org/10.1038/nbt.3238
-- 
--   Myers EW, et al.
--   A Whole-Genome Assembly of Drosophila.
--   Science. 2000 Mar 24;287(5461):2196-204.
--   http://doi.org/10.1126/science.287.5461.2196
-- 
-- Corrected read consensus sequences are generated using an algorithm derived from FALCON-sense:
--   Chin CS, et al.
--   Phased diploid genome assembly with single-molecule real-time sequencing.
--   Nat Methods. 2016 Dec;13(12):1050-1054.
--   http://doi.org/10.1038/nmeth.4035
-- 
-- Contig consensus sequences are generated using an algorithm derived from pbdagcon:
--   Chin CS, et al.
--   Nonhybrid, finished microbial genome assemblies from long-read SMRT sequencing data.
--   Nat Methods. 2013 Jun;10(6):563-9
--   http://doi.org/10.1038/nmeth.2474
-- 
-- CONFIGURE CANU
--
-- Detected Java(TM) Runtime Environment '1.8.0_262' (from 'java') with -d64 support.
--
-- WARNING:
-- WARNING:  Failed to run gnuplot using command 'gnuplot'.
-- WARNING:  Plots will be disabled.
-- WARNING:
--
-- Detected 36 CPUs and 1008 gigabytes of memory.
-- Detected Slurm with 'sinfo' binary in /usr/bin/sinfo.
-- Grid engine and staging disabled per useGrid=false option.
--
--                                (tag)Concurrency
--                         (tag)Threads          |
--                (tag)Memory         |          |
--        (tag)             |         |          |       total usage      algorithm
--        -------  ----------  --------   --------  --------------------  -----------------------------
-- Local: meryl     24.000 GB    6 CPUs x   6 jobs   144.000 GB  36 CPUs  (k-mer counting)
-- Local: hap       12.000 GB   18 CPUs x   2 jobs    24.000 GB  36 CPUs  (read-to-haplotype assignment)
-- Local: cormhap   32.000 GB   12 CPUs x   3 jobs    96.000 GB  36 CPUs  (overlap detection with mhap)
-- Local: obtovl    16.000 GB   12 CPUs x   3 jobs    48.000 GB  36 CPUs  (overlap detection)
-- Local: utgovl    16.000 GB   12 CPUs x   3 jobs    48.000 GB  36 CPUs  (overlap detection)
-- Local: cor       16.000 GB    4 CPUs x   9 jobs   144.000 GB  36 CPUs  (read correction)
-- Local: ovb        4.000 GB    1 CPU  x  36 jobs   144.000 GB  36 CPUs  (overlap store bucketizer)
-- Local: ovs       16.000 GB    1 CPU  x  36 jobs   576.000 GB  36 CPUs  (overlap store sorting)
-- Local: red       32.000 GB    6 CPUs x   6 jobs   192.000 GB  36 CPUs  (read error detection)
-- Local: oea        8.000 GB    1 CPU  x  36 jobs   288.000 GB  36 CPUs  (overlap error adjustment)
-- Local: bat      256.000 GB   16 CPUs x   1 job    256.000 GB  16 CPUs  (contig construction with bogart)
-- Local: cns        -.--- GB    8 CPUs x   - jobs     -.--- GB   - CPUs  (consensus)
--
-- Found trimmed corrected Nanopore reads in the input files.
--
-- Generating assembly 'Canu_AllData_1000' in '/projects/EBP/CBC/pretoria/AC_SP644/Canu_assembly':
--    - assemble corrected and trimmed reads.
--
-- Parameters:
--
--  genomeSize        680000000
--
--  Overlap Generation Limits:
--    corOvlErrorRate 0.3200 ( 32.00%)
--    obtOvlErrorRate 0.1200 ( 12.00%)
--    utgOvlErrorRate 0.1200 ( 12.00%)
--
--  Overlap Processing Limits:
--    corErrorRate    0.5000 ( 50.00%)
--    obtErrorRate    0.1200 ( 12.00%)
--    utgErrorRate    0.1200 ( 12.00%)
--    cnsErrorRate    0.2000 ( 20.00%)
--
--
-- BEGIN ASSEMBLY
--
----------------------------------------
-- Starting command on Thu Jul 29 08:29:10 2021 with 6190.83 GB free disk space

    cd .
    ./Canu_AllData_1000.seqStore.sh \
    > ./Canu_AllData_1000.seqStore.err 2>&1

-- Finished on Thu Jul 29 08:32:14 2021 (184 seconds) with 6184.638 GB free disk space
----------------------------------------
----------------------------------------
-- Starting command on Thu Jul 29 08:32:14 2021 with 6184.638 GB free disk space

    cd .
    /isg/shared/apps/canu/2.1.1/bin/sqStoreDumpMetaData \
      -S ./Canu_AllData_1000.seqStore \
      -corrected -trimmed \
      -histogram \
      -lengths \
    > ./Canu_AllData_1000.seqStore/readlengths-utg.dat \
    2> ./Canu_AllData_1000.seqStore/readlengths-utg.err 
    

-- Finished on Thu Jul 29 08:32:14 2021 (like a bat out of hell) with 6184.638 GB free disk space
----------------------------------------
--
-- WARNING: gnuplot failed.
--
----------------------------------------
----------------------------------------
-- Starting command on Thu Jul 29 08:32:14 2021 with 6184.638 GB free disk space

    cd .
    /isg/shared/apps/canu/2.1.1/bin/sqStoreDumpMetaData \
      -S ./Canu_AllData_1000.seqStore \
      -corrected -trimmed \
      -histogram \
    > ./Canu_AllData_1000.seqStore/readlengths-utg.txt \
    2> ./Canu_AllData_1000.seqStore/readlengths-utg.err 
    

-- Finished on Thu Jul 29 08:32:14 2021 (lickety-split) with 6184.638 GB free disk space
----------------------------------------
--
-- In sequence store './Canu_AllData_1000.seqStore':
--   Found 448270 reads.
--   Found 20401859798 bases (30 times coverage).
--    
--    G=20401859798                      sum of  ||               length     num
--    NG         length     index       lengths  ||                range    seqs
--    ----- ------------ --------- ------------  ||  ------------------- -------
--    00010        55401     32930   2040216379  ||      39104-41390      129565|---------------------------------------------------------------
--    00020        50662     71632   4080394134  ||      41391-43677       94975|-----------------------------------------------
--    00030        47900    113120   6120563797  ||      43678-45964       68657|----------------------------------
--    00040        45900    156671   8160782247  ||      45965-48251       48264|------------------------
--    00050        44323    201927  10200949671  ||      48252-50538       33686|-----------------
--    00060        43007    248673  12241123279  ||      50539-52825       23051|------------
--    00070        41860    296769  14281316517  ||      52826-55112       15556|--------
--    00080        40852    346119  16321522919  ||      55113-57399       10571|------
--    00090        39936    396638  18361675218  ||      57400-59686        7241|----
--    00100        39104    448269  20401859798  ||      59687-61973        5077|---
--    001.000x              448270  20401859798  ||      61974-64260        3477|--
--                                               ||      64261-66547        2353|--
--                                               ||      66548-68834        1667|-
--                                               ||      68835-71121        1146|-
--                                               ||      71122-73408         784|-
--                                               ||      73409-75695         563|-
--                                               ||      75696-77982         402|-
--                                               ||      77983-80269         311|-
--                                               ||      80270-82556         240|-
--                                               ||      82557-84843         150|-
--                                               ||      84844-87130         145|-
--                                               ||      87131-89417          83|-
--                                               ||      89418-91704          60|-
--                                               ||      91705-93991          58|-
--                                               ||      93992-96278          42|-
--                                               ||      96279-98565          29|-
--                                               ||      98566-100852         16|-
--                                               ||     100853-103139         19|-
--                                               ||     103140-105426         16|-
--                                               ||     105427-107713         15|-
--                                               ||     107714-110000         12|-
--                                               ||     110001-112287          8|-
--                                               ||     112288-114574          7|-
--                                               ||     114575-116861          4|-
--                                               ||     116862-119148          7|-
--                                               ||     119149-121435          3|-
--                                               ||     121436-123722          1|-
--                                               ||     123723-126009          1|-
--                                               ||     126010-128296          2|-
--                                               ||     128297-130583          0|
--                                               ||     130584-132870          2|-
--                                               ||     132871-135157          2|-
--                                               ||     135158-137444          0|
--                                               ||     137445-139731          0|
--                                               ||     139732-142018          0|
--                                               ||     142019-144305          1|-
--                                               ||     144306-146592          0|
--                                               ||     146593-148879          0|
--                                               ||     148880-151166          0|
--                                               ||     151167-153453          1|-
--    
----------------------------------------
-- Starting command on Thu Jul 29 08:32:14 2021 with 6184.638 GB free disk space

    cd unitigging/0-mercounts
    ./meryl-configure.sh \
    > ./meryl-configure.err 2>&1

-- Finished on Thu Jul 29 08:32:15 2021 (one second) with 6184.638 GB free disk space
----------------------------------------
--  segments   memory batches
--  -------- -------- -------
--        01 24.00 GB       3
--        02 24.00 GB       2
--        04 16.79 GB       1
--        06 12.05 GB       1
--        08  9.05 GB       1
--        12  6.05 GB       1
--        16  5.07 GB       1
--        20  4.07 GB       1
--        24  3.05 GB       1
--        32  2.57 GB       1
--        40  2.07 GB       1
--        48  1.69 GB       1
--        56  1.50 GB       1
--        64  1.32 GB       1
--        96  0.88 GB       1
--
--  For 448270 reads with 20401859798 bases, limit to 204 batches.
--  Will count kmers using 04 jobs, each using 18 GB and 6 threads.
--
-- Finished stage 'merylConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'meryl' concurrent execution on Thu Jul 29 08:32:15 2021 with 6184.638 GB free disk space (4 processes; 6 concurrently)

    cd unitigging/0-mercounts
    ./meryl-count.sh 1 > ./meryl-count.000001.out 2>&1
    ./meryl-count.sh 2 > ./meryl-count.000002.out 2>&1
    ./meryl-count.sh 3 > ./meryl-count.000003.out 2>&1
    ./meryl-count.sh 4 > ./meryl-count.000004.out 2>&1

-- Finished on Thu Jul 29 08:41:06 2021 (531 seconds) with 6155.556 GB free disk space
----------------------------------------
-- Found 4 Kmer counting (meryl) outputs.
-- Finished stage 'utg-merylCountCheck', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'meryl' concurrent execution on Thu Jul 29 08:41:06 2021 with 6155.556 GB free disk space (1 processes; 6 concurrently)

    cd unitigging/0-mercounts
    ./meryl-process.sh 1 > ./meryl-process.000001.out 2>&1

-- Finished on Thu Jul 29 08:42:13 2021 (67 seconds) with 6152.962 GB free disk space
----------------------------------------
-- Meryl finished successfully.  Kmer frequency histogram:
--
-- WARNING: gnuplot failed.
--
----------------------------------------
--
--  22-mers                                                                                           Fraction
--    Occurrences   NumMers                                                                         Unique Total
--       1-     1         0                                                                        0.0000 0.0000
--       2-     2 118833745 ********************************************************************** 0.1430 0.0121
--       3-     4  93560810 *******************************************************                0.2116 0.0208
--       5-     7  71612746 ******************************************                             0.2887 0.0352
--       8-    11  88973536 ****************************************************                   0.3670 0.0582
--      12-    16 109831339 ****************************************************************       0.4780 0.1077
--      17-    22  88259541 ***************************************************                    0.6013 0.1849
--      23-    29 111854694 *****************************************************************      0.7051 0.2748
--      30-    37  80430470 ***********************************************                        0.8399 0.4283
--      38-    46  23690301 *************                                                          0.9241 0.5483
--      47-    56  11340254 ******                                                                 0.9488 0.5923
--      57-    67   8198804 ****                                                                   0.9619 0.6211
--      68-    79   5378011 ***                                                                    0.9714 0.6461
--      80-    92   3948938 **                                                                     0.9776 0.6656
--      93-   106   2778332 *                                                                      0.9822 0.6825
--     107-   121   1975420 *                                                                      0.9855 0.6961
--     122-   137   1501488                                                                        0.9878 0.7074
--     138-   154   1168873                                                                        0.9895 0.7171
--     155-   172    937686                                                                        0.9909 0.7257
--     173-   191    766434                                                                        0.9920 0.7334
--     192-   211    632777                                                                        0.9930 0.7404
--     212-   232    528191                                                                        0.9937 0.7468
--     233-   254    447475                                                                        0.9943 0.7527
--     255-   277    384078                                                                        0.9949 0.7582
--     278-   301    331903                                                                        0.9953 0.7634
--     302-   326    291913                                                                        0.9957 0.7683
--     327-   352    256115                                                                        0.9961 0.7729
--     353-   379    228227                                                                        0.9964 0.7773
--     380-   407    201654                                                                        0.9967 0.7815
--     408-   436    181645                                                                        0.9969 0.7855
--     437-   466    162398                                                                        0.9971 0.7894
--     467-   497    146646                                                                        0.9973 0.7931
--     498-   529    132483                                                                        0.9975 0.7967
--     530-   562    121065                                                                        0.9976 0.8002
--     563-   596    109293                                                                        0.9978 0.8035
--     597-   631     99856                                                                        0.9979 0.8067
--     632-   667     90662                                                                        0.9980 0.8098
--     668-   704     82603                                                                        0.9981 0.8128
--     705-   742     76209                                                                        0.9982 0.8157
--     743-   781     70841                                                                        0.9983 0.8185
--     782-   821     66248                                                                        0.9984 0.8212
--
--           0 (max occurrences)
-- 19671072496 (total mers, non-unique)
--   830932686 (distinct mers, non-unique)
--           0 (unique mers)
-- Finished stage 'meryl-process', reset canuIteration.
--
-- Removing meryl database 'unitigging/0-mercounts/Canu_AllData_1000.ms22'.
--
-- OVERLAPPER (normal) (assembly) erate=0.12
--
----------------------------------------
-- Starting command on Thu Jul 29 08:42:15 2021 with 6152.962 GB free disk space

    cd unitigging/1-overlapper
    /isg/shared/apps/canu/2.1.1/bin/overlapInCorePartition \
     -S  ../../Canu_AllData_1000.seqStore \
     -hl 320000000 \
     -rl 15000000000 \
     -ol 500 \
     -o  ./Canu_AllData_1000.partition \
    > ./Canu_AllData_1000.partition.err 2>&1

-- Finished on Thu Jul 29 08:42:15 2021 (fast as lightning) with 6152.962 GB free disk space
----------------------------------------
--
-- Configured 82 overlapInCore jobs.
-- Finished stage 'utg-overlapConfigure', reset canuIteration.
--
-- Running jobs.  First attempt out of 2.
----------------------------------------
-- Starting 'utgovl' concurrent execution on Thu Jul 29 08:42:15 2021 with 6152.962 GB free disk space (82 processes; 3 concurrently)

    cd unitigging/1-overlapper
    ./overlap.sh 1 > ./overlap.000001.out 2>&1
    ./overlap.sh 2 > ./overlap.000002.out 2>&1
    ./overlap.sh 3 > ./overlap.000003.out 2>&1
    ./overlap.sh 4 > ./overlap.000004.out 2>&1
    ./overlap.sh 5 > ./overlap.000005.out 2>&1
    ./overlap.sh 6 > ./overlap.000006.out 2>&1
    ./overlap.sh 7 > ./overlap.000007.out 2>&1
    ./overlap.sh 8 > ./overlap.000008.out 2>&1
    ./overlap.sh 9 > ./overlap.000009.out 2>&1
    ./overlap.sh 10 > ./overlap.000010.out 2>&1
    ./overlap.sh 11 > ./overlap.000011.out 2>&1
    ./overlap.sh 12 > ./overlap.000012.out 2>&1
    ./overlap.sh 13 > ./overlap.000013.out 2>&1
    ./overlap.sh 14 > ./overlap.000014.out 2>&1
    ./overlap.sh 15 > ./overlap.000015.out 2>&1
    ./overlap.sh 16 > ./overlap.000016.out 2>&1
    ./overlap.sh 17 > ./overlap.000017.out 2>&1
    ./overlap.sh 18 > ./overlap.000018.out 2>&1
    ./overlap.sh 19 > ./overlap.000019.out 2>&1
    ./overlap.sh 20 > ./overlap.000020.out 2>&1
    ./overlap.sh 21 > ./overlap.000021.out 2>&1
    ./overlap.sh 22 > ./overlap.000022.out 2>&1
    ./overlap.sh 23 > ./overlap.000023.out 2>&1
    ./overlap.sh 24 > ./overlap.000024.out 2>&1
    ./overlap.sh 25 > ./overlap.000025.out 2>&1
    ./overlap.sh 26 > ./overlap.000026.out 2>&1
    ./overlap.sh 27 > ./overlap.000027.out 2>&1
    ./overlap.sh 28 > ./overlap.000028.out 2>&1
    ./overlap.sh 29 > ./overlap.000029.out 2>&1
    ./overlap.sh 30 > ./overlap.000030.out 2>&1
    ./overlap.sh 31 > ./overlap.000031.out 2>&1
    ./overlap.sh 32 > ./overlap.000032.out 2>&1
    ./overlap.sh 33 > ./overlap.000033.out 2>&1
    ./overlap.sh 34 > ./overlap.000034.out 2>&1
    ./overlap.sh 35 > ./overlap.000035.out 2>&1
```

Thank you for the help.

**Comment by skoren** [comment]
Canu is primarily designed to use a grid, running an almost 1gb genome with ONT data on only 28 cores will take a while, as you've found out. How did you obtain the corrected reads? If they're not from Canu, especially if they're corrected using short-read data, I wouldn't recommend using them. Instead you could try the experimental uncorrected assembly for ONT reads (assuming you have used a recent base caller like guppy v5+) as documented here: https://canu.readthedocs.io/en/latest/quick-start.html#correct-trim-and-assemble-manually.

Hitting the 21-day limit isn't going to lose all your compute because canu can resume a run where it left off so you could still run the original command (correcting/trimming/assembling) and when it runs out of time, re-launch the same command to let it continue. There are a few options suggested on the FAQ to speed up the run: https://canu.readthedocs.io/en/latest/faq.html#my-assembly-is-running-out-of-space-is-too-slow which you can try along with lowering the default corrected error rate from 12 to 8% via `correctedErrorRate=0.08` which should be sufficient for newer ONT data.

**Comment by Mondlii** [comment]
Hi Sergey Koren, Thank you for the feedback.

I used the corrected reads I had from another assembly run (Necat assembly). But I retried running without correcting the reads and running the whole process and after subsampling from the 200x coverage to about 80x (with hopes of correcting using the 200x after assembly) the run seems to moving at a reasonable pace and has reached the trimming stage after day 3 of running. Also this might also be boosted by the different cluster I am using with 750GB memory (even at slightly fewer cores than the 28 I mentioned before). 

Thank you again for the response and assistance. I will make use of the docs and additional parameters to test out different run conditions and commands.
